### PR TITLE
os/bluestore: add bluestore_prefer_wal_size option

### PR DESCRIPTION
**Pull request by liewegas**
Add option to prefer a WAL write if the write is below a size threshold,
even if we could avoid it.  This lets you trade some write-amp (by
journaling data to rocksdb) for latency in cases where the WAL device is
much faster than the main device.

This affects:

 - writes to new extents locations below min_alloc_size
 - writes to unallocated space below min_alloc_size
 - "big" writes above min_alloc_size that are below the prefer_wal_size
   threshold.

Note that it's applied to individual blobs, not the entirety of the write,
so if your have a larger write torn into two pieces/blobs that are below
the threshold then they will both go through the wal.

Also, do not wake kv sync thread to retire wal events; do that lazily or on flush() (~50 -> ~100 IOPS)

### CODE REVIEW

**Comment by ifed01** [comment]
need rebase

**Comment by varadakari** [comment]
some getattr failures in the make check.

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -2801,6 +2803,10 @@ void BlueStore::handle_conf_change(const struct md_config_t *conf,
       changed.count("bluestore_compression_max_blob_size")) {
     _set_compression();
   }
+  if (changed.count("bluestore_prefer_wal_size") ||
+      changed.count("bluestore_max_alloc_size")) {
+    _set_alloc_sizes();
```
Comment (markhpc) [inline src/os/bluestore/BlueStore.cc]: There doesn't appear to be any check to make sure that handle_conf_change doesn't happen before BlueStore::mount() (or mkfs/fsck, ie anything that opens _open_bdev).  As a result, bdev can be empty when _set_alloc_sizes() is called.

Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: fixes

**Comment by liewegas** [comment]
rebased, and updated to avoid spurious kv thread wakes+commits just to retire wal records.  now 2x as fast as before when qd=1

**Comment by liewegas** [comment]
Now 4x faster (400 IOPS) with the wal aio dispatch batching!

**Comment by liewegas** [comment]
Okay, there are more fundamental problems with the wal writes:

#. we don't prevent multiple in-flight wal writes to the same block (the block layer is unordered)
#. we don't prevent a block with an in-flight wal write from being deallocated (except in the most trivial cases).

I also have a bunch of cleanups to flush() and friends, but I'll wrap those into a larger bugfix.  In the meantime, making this change minimal to address just the prefer_wal_size case!

**Comment by liewegas** [comment]
...aaand dropping the no-wakeup commit since it can also easily deadlock.  I'll include it with the flush cleanup later.

**Pull request merged**